pkg/controller/che/che_controller.go

```diff
@@ -1123,3 +1089,51 @@ func isTrustedBundleConfigMap(mgr manager.Manager, obj handler.MapObject) (bool,
 		},
 	}
 }
+
+func (r *ReconcileChe) autoEnableOAuth(cr *orgv1.CheCluster, request reconcile.Request, isOpenShift4 bool) (reconcile.Result, error) {
+	var message, reason string
+	if isOpenShift4 {
+		oauthv1 := &oauthv1.OAuth{}
+		if err := r.nonCachedClient.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, oauthv1); err != nil {
+			getOAuthV1ErrMsg := failedUnableToGetOAuth + " Cause: " + err.Error()
+			logrus.Errorf(getOAuthV1ErrMsg)
+			message = getOAuthV1ErrMsg
+			reason = failedNoOpenshiftUserReason
+			cr.Spec.Auth.OpenShiftoAuth = util.NewBoolPointer(false)
+		} else {
+			cr.Spec.Auth.OpenShiftoAuth = util.NewBoolPointer(len(oauthv1.Spec.IdentityProviders) >= 1)
+			if !*cr.Spec.Auth.OpenShiftoAuth {
+				logrus.Warn(warningNoIdentityProvidersMessage, " ", howToAddIdentityProviderLinkOS4)
+			}
+		}
+		// openshift 3
+	} else {
+		users := &userv1.UserList{}
+		listOptions := &client.ListOptions{}
+		if err := r.nonCachedClient.List(context.TODO(), users, listOptions); err != nil {
+			getUsersErrMsg := failedUnableToGetOpenshiftUsers + " Cause: " + err.Error()
+			logrus.Errorf(getUsersErrMsg)
+			message = getUsersErrMsg
+			reason = failedNoOpenshiftUserReason
+			cr.Spec.Auth.OpenShiftoAuth = util.NewBoolPointer(false)
+		}
+
+		cr.Spec.Auth.OpenShiftoAuth = util.NewBoolPointer(len(users.Items) >= 1)
+		if !*cr.Spec.Auth.OpenShiftoAuth {
+			logrus.Warn(warningNoRealUsersMessage, " ", howToConfigureOAuthLinkOS3)
+		}
+	}
+
+	oauth := *cr.Spec.Auth.OpenShiftoAuth
+	if err := r.UpdateCheCRSpec(cr, "OpenShiftoAuth", strconv.FormatBool(oauth)); err != nil {
+		return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 1}, err
+	}
```

Comment:
Compare auth and cr.Spec.Auth.OpenShiftoAuth, if they are different then update CR

Reply:
Do we really need this? `autoEnableOAuth` method will be executed only if `openshiftOAuth` has empty value(nil for golang code, user didn't set up obviously true or false). Then `auto-detect` code set up true/false value and this true/false is always not equal to initial nil.